### PR TITLE
Add Ubuntu 24.04 LTS for VMware vSphere

### PR DESCRIPTION
**Pull request by sriramandev**
## Change description
Adds support for Ubuntu 24.04 LTS for VMware vSphere

## 
Related issues

- Refs: https://github.com/kubernetes-sigs/image-builder/issues/1406
- See: https://github.com/kubernetes-sigs/image-builder/pull/1003

## Additional context

- ubuntu 24.04 now uses Python 3.12. With Python 3.10 distutils package was deprecated and with python 3.12 has been removed
- With Ubuntu 24.04, the location of package source management has moved from location /etc/apt/sources.list to    /etc/apt/sources.list.d/ubuntu.sources and the format now adheres to deb822 format

/cc @kkeshavamurthy @mboersma

### CODE REVIEW

**Comment by sriramandev** [comment]
@mboersma The `pull-lint` job failures are not related to the code changes wrt to this PR... how do we proceed on that?

**Comment by AverageMarcus** [comment]
Two of the failures look like they might be related to this PR:
```
yaml[trailing-spaces]: Trailing spaces
ansible/roles/node/defaults/main.yml:99
yaml[trailing-spaces]: Trailing spaces
ansible/roles/node/meta/main.yml:36
```
(See https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_image-builder/1459/pull-lint/1786044902249861120#1:build-log.txt%3A1158)

**Comment by sriramandev** [comment]
> Two of the failures look like they might be related to this PR:
> 
> ```
> yaml[trailing-spaces]: Trailing spaces
> ansible/roles/node/defaults/main.yml:99
> yaml[trailing-spaces]: Trailing spaces
> ansible/roles/node/meta/main.yml:36
> ```
> 
> (See https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_image-builder/1459/pull-lint/1786044902249861120#1:build-log.txt%3A1158)

Fixed the linter issues mentioned above... Thanks..

**Comment by AverageMarcus** [comment]
It looks like something might not be right with the 24.04 builds. Based on the log from that last run it doesn't get past the ssh stage: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_image-builder/1459/pull-ova-all/1786249412729638912/artifacts/ubuntu-2404.log

**Comment by sriramandev** [comment]
> It looks like something might not be right with the 24.04 builds. Based on the log from that last run it doesn't get past the ssh stage: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_image-builder/1459/pull-ova-all/1786249412729638912/artifacts/ubuntu-2404.log

How do you debug this without VC details? Any help is appreciated...

**Comment by AverageMarcus** [comment]
I'm not sure about VSphere myself 😞 Maybe @kkeshavamurthy can offer some suggestion?

File: values.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (AverageMarcus) [inline values.json:1]: @sriramandev I'm assuming you didn't mean to commit these files in?

Comment (sriramandev) [reply values.json]: Yes... my mistake...these should not have got in

**Comment by rajaskakodkar** [comment]
Currently the failure in the `pull-ova-all` job is
```
ack/ensure-ovftool.sh
/home/prow/go/src/sigs.k8s.io/image-builder/images/capi/.local/bin/packer init packer/config.pkr.hcl
/home/prow/go/src/sigs.k8s.io/image-builder/images/capi/.local/bin/packer build -var-file="/home/prow/go/src/sigs.k8s.io/image-builder/images/capi/packer/config/kubernetes.json"  -var-file="/home/prow/go/src/sigs.k8s.io/image-builder/images/capi/packer/config/cni.json"  -var-file="/home/prow/go/src/sigs.k8s.io/image-builder/images/capi/packer/config/containerd.json"  -var-file="/home/prow/go/src/sigs.k8s.io/image-builder/images/capi/packer/config/wasm-shims.json"  -var-file="/home/prow/go/src/sigs.k8s.io/image-builder/images/capi/packer/config/ansible-args.json"  -var-file="/home/prow/go/src/sigs.k8s.io/image-builder/images/capi/packer/config/goss-args.json"  -var-file="/home/prow/go/src/sigs.k8s.io/image-builder/images/capi/packer/config/common.json"  -var-file="/home/prow/go/src/sigs.k8s.io/image-builder/images/capi/packer/config/additional_components.json"  -color=true  -var-file="packer/ova/packer-common.json" -var-file="/home/prow/go/src/sigs.k8s.io/image-builder/images/capi/packer/ova/ubuntu-2404.json" -var-file="packer/ova/vsphere.json"  -except=local -only=vsphere-iso -var-file="/home/prow/go/src/sigs.k8s.io/image-builder/images/capi/ci-ubuntu-2404.json"  -var-file="/home/prow/go/src/sigs.k8s.io/image-builder/images/capi/scripts/ci-disable-goss-inspect.json"  -only=vsphere packer/ova/packer-node.json
Warning: Bundled plugins used

This template relies on the use of plugins bundled into the Packer binary.
The practice of bundling external plugins into Packer will be removed in an
upcoming version.

To remove this warning and ensure builds keep working you can install these
external plugins with the 'packer plugins install' command

* packer plugins install github.com/hashicorp/vmware
* packer plugins install github.com/hashicorp/vsphere

Alternatively, if you upgrade your templates to HCL2, you can use 'packer init'
with a 'required_plugins' block to automatically install external plugins.

You can try HCL2 by running 'packer hcl2_upgrade
/home/prow/go/src/sigs.k8s.io/image-builder/images/capi/packer/ova/packer-node.json'


[1;32mvsphere-iso.vsphere: output will be in this color.[0m

[1;32m==> vsphere-iso.vsphere: File /root/.cache/packer/bf6042c64ad2c3a07e3e59346064d207b5908877.iso already uploaded; continuing[0m
[1;32m==> vsphere-iso.vsphere: File [WorkloadDatastore] packer_cache//bf6042c64ad2c3a07e3e59346064d207b5908877.iso already exists; skipping upload.[0m
[1;32m==> vsphere-iso.vsphere: Creating VM...[0m
[1;32m==> vsphere-iso.vsphere: Customizing hardware...[0m
[1;32m==> vsphere-iso.vsphere: Mounting ISO images...[0m
[1;32m==> vsphere-iso.vsphere: Adding configuration parameters...[0m
[1;32m==> vsphere-iso.vsphere: Creating floppy disk...[0m
[0;32m    vsphere-iso.vsphere: Copying files flatly from floppy_files[0m
[0;32m    vsphere-iso.vsphere: Done copying files from floppy_files[0m
[0;32m    vsphere-iso.vsphere: Collecting paths from floppy_dirs[0m
[0;32m    vsphere-iso.vsphere: Resulting paths from floppy_dirs : [./packer/ova/linux/ubuntu/http/][0m
[0;32m    vsphere-iso.vsphere: Recursively copying : ./packer/ova/linux/ubuntu/http/[0m
[0;32m    vsphere-iso.vsphere: Done copying paths from floppy_dirs[0m
[0;32m    vsphere-iso.vsphere: Copying files from floppy_content[0m
[0;32m    vsphere-iso.vsphere: Done copying files from floppy_content[0m
[1;32m==> vsphere-iso.vsphere: Uploading created floppy image[0m
[1;32m==> vsphere-iso.vsphere: Adding generated Floppy...[0m
[1;32m==> vsphere-iso.vsphere: Starting HTTP server on port 8486[0m
[1;32m==> vsphere-iso.vsphere: Set boot order temporary...[0m
[1;32m==> vsphere-iso.vsphere: Power on VM...[0m
[1;32m==> vsphere-iso.vsphere: Waiting 10s for boot...[0m
[1;32m==> vsphere-iso.vsphere: HTTP server is working at http://10.60.145.16:8486/[0m
[1;32m==> vsphere-iso.vsphere: Typing boot command...[0m
[1;32m==> vsphere-iso.vsphere: Waiting for IP...[0m
[1;32m==> vsphere-iso.vsphere: IP address: 192.168.24.71[0m
[1;32m==> vsphere-iso.vsphere: Using SSH communicator to connect: 192.168.24.71[0m
[1;32m==> vsphere-iso.vsphere: Waiting for SSH to become available...[0m
```
where packer is waiting for packer to be available.

@DimpleRajaVamsi does this look similar to what you faced earlier in https://github.com/kubernetes-sigs/image-builder/issues/1236? 

https://github.com/kubernetes-sigs/image-builder/pull/1237 mentions that timeout is a workaround and not a fix, do you have any ideas on what would be a fix here?

File: images/capi/scripts/ci-ova.sh [hunk 2]
```diff
@@ -22,7 +22,7 @@ CAPI_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
 cd "${CAPI_ROOT}" || exit 1
 
 export ARTIFACTS="${ARTIFACTS:-${PWD}/_artifacts}"
-TARGETS=("ubuntu-2004" "ubuntu-2204" "photon-3" "photon-4" "photon-5" "rockylinux-8" "flatcar")
+TARGETS=("ubuntu-2004" "ubuntu-2204" "ubuntu-2404" "photon-3" "photon-4" "photon-5" "rockylinux-8" "flatcar")
```
Comment (rajaskakodkar) [inline images/capi/scripts/ci-ova.sh:26]: @sriramandev - I see that this packer config https://github.com/kubernetes-sigs/image-builder/pull/1003/files#diff-92590d094272c3e33208dbb6a15aff2a1f1e9d866b7411a2c38271e57f5b2728R119-R125 is not included in this PR. Could that be the issue?

File: images/capi/packer/ova/ubuntu-2404-efi.json [hunk 3]
```diff
@@ -0,0 +1,18 @@
+{
+  "boot_command_prefix": "c<wait>linux /casper/vmlinuz ipv6.disable={{ user `boot_disable_ipv6` }} --- autoinstall ds='nocloud-net;s=http://{{ .HTTPIP }}:{{ .HTTPPort }}/24.04.efi/'<enter><wait>initrd /casper/initrd<enter><wait>boot<enter>",
```
Comment (chrischdi) [inline images/capi/packer/ova/ubuntu-2404-efi.json:2]:
```suggestion
  "boot_command_prefix": "c<wait>linux /casper/vmlinuz ipv6.disable={{ user `boot_disable_ipv6` }} --- autoinstall ds='nocloud'<enter><wait>initrd /casper/initrd<enter><wait>boot<enter>",
```
Maybe this helps (also at the other).

We're trying to read the data from a floppy, not via http. The VM is not able to reach the processes we are running here.

Comment (rajaskakodkar) [reply images/capi/packer/ova/ubuntu-2404-efi.json:2]: Oh that is interesting. Worth trying this out

**Comment by chrischdi** [comment]
Looks like it works now to ssh to the vm:

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_image-builder/1459/pull-ova-all/1815643225218617344/artifacts/ubuntu-2404.log

Just fails somewhere in ansible 🎉

**Comment by rajaskakodkar** [comment]
The fix was to try CDROM boot instead of http to pass cloudinit data. Looks like the goss failure is a flake.

File: images/capi/packer/ova/packer-node.json [hunk 4]
```diff
@@ -200,6 +200,10 @@
       ],
       "boot_wait": "{{user `boot_wait`}}",
       "cdrom_type": "{{user `cdrom_type`}}",
+      "cd_files": [
+        "{{user `cd_files`}}"
+      ],
```
Comment (chrischdi) [inline images/capi/packer/ova/packer-node.json:205]: I think this needs to not always be set, e.g. when ```user `cd_files```` is empty, it should not set an array here with a single empty string.

Comment (sriramandev) [reply images/capi/packer/ova/packer-node.json]: Fixed this in the latest force update. Thanks

**Comment by sriramandev** [comment]
We have moved to using CD mounts to provide the cloud-init boot files. For this we have introduced a new package `xorriso` in the Dockerfile which gets used by the packer to create the ISO image out of the boot file. Similarly for CI, ci-ova.sh has been updated to include the same.
Along with ubuntu-24.04, we have also moved ubuntu-22.04,photon-5 to use CD based cloud-init. Ubuntu-20.04 continues to use floppy drive.

**Comment by sriramandev** [comment]
@chrischdi / @mboersma / @AverageMarcus Can any of you please review the changes and merge them. It would be great if we could get this in before the disabling of OVA CI.

File: images/capi/packer/ova/ubuntu-2204.json [hunk 5]
```diff
@@ -1,12 +1,13 @@
 {
-  "boot_command_prefix": "c<wait>linux /casper/vmlinuz ipv6.disable={{ user `boot_disable_ipv6` }} --- autoinstall ds='nocloud-net;s=http://{{ .HTTPIP }}:{{ .HTTPPort }}/22.04/'<enter><wait>initrd /casper/initrd<enter><wait>boot<enter>",
+  "boot_command_prefix": "c<wait>linux /casper/vmlinuz ipv6.disable={{ user `boot_disable_ipv6` }} --- autoinstall ds='nocloud;'<enter><wait>initrd /casper/initrd<enter><wait>boot<enter><wait300>",
```
Comment (chrischdi) [inline images/capi/packer/ova/ubuntu-2204.json:2]:
```suggestion
  "boot_command_prefix": "c<wait>linux /casper/vmlinuz ipv6.disable={{ user `boot_disable_ipv6` }} --- autoinstall ds='nocloud;'<enter><wait>initrd /casper/initrd<enter><wait>boot<enter>",
```

What do we need this wait for?

**Review by mboersma — APPROVED** [review]
/lgtm
/assign @drew-viles @AverageMarcus 
/retest-required

Azure failure looks like a build flake.

**Comment by drew-viles** [comment]
All looks good to me. However, I'm not a vsphere expert by any means. Based on the conversation it seems everyone is happy with the progress made in this PR though so I will

/approve

**Comment by rajaskakodkar** [comment]
> We have moved to using CD mounts to provide the cloud-init boot files. For this we have introduced a new package xorriso in the Dockerfile which gets used by the packer to create the ISO image out of the boot file. Similarly for CI, ci-ova.sh has been updated to include the same.
Along with ubuntu-24.04, we have also moved ubuntu-22.04,photon-5 to use CD based cloud-init. Ubuntu-20.04 continues to use floppy drive.

For more context, it looks like we were hitting this race condition in cloudinit with http - https://github.com/canonical/cloud-init/issues/4044

**Pull request merged**